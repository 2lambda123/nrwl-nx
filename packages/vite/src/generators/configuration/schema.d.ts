export interface Schema {
  uiFramework: 'react' | 'none';
  project: string;
  newProject?: boolean;
}
