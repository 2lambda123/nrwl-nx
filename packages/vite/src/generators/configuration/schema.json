{
  "cli": "nx",
  "title": "Add Vite Configuration to an application.",
  "description": "Add Vite Configuration to an application.",
  "$id": "configure-vite-app",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "aliases": ["name", "projectName"],
      "x-dropdown": "project",
      "x-prompt": "What is the name of the project to set up a webpack for?"
    },
    "includeLib": {
      "type": "boolean",
      "description": "Add a library build option and skip the server option.",
      "default": false,
      "x-prompt": "Does this project contain a buildable library?"
    },
    "uiFramework": {
      "type": "string",
      "description": "UI Framework to use for Vite.",
      "enum": ["react", "none"],
      "default": "none",
      "x-prompt": "What UI framework plugin should Vite use?"
    },
    "newProject": {
      "type": "boolean",
      "description": "Is this a new project?",
      "default": false
    }
  },
  "examplesFile": "../../../docs/configuration-examples.md"
}
