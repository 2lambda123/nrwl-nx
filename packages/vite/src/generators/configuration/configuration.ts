import {
  convertNxGenerator,
  formatFiles,
  GeneratorCallback,
  readProjectConfiguration,
  Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import {
  findExistingTargets,
  addOrChangeBuildTarget,
  addOrChangeServeTarget,
  editTsConfig,
  moveAndEditIndexHtml,
  writeViteConfig,
} from '../../utils/generator-utils';

import initGenerator from '../init/init';
import vitestGenerator from '../vitest/vitest-generator';
import { Schema } from './schema';

export async function viteConfigurationGenerator(tree: Tree, schema: Schema) {
  const tasks: GeneratorCallback[] = [];

  const { targets, projectType } = readProjectConfiguration(
    tree,
    schema.project
  );
  let buildTargetName = 'build';
  let serveTargetName = 'serve';

  schema.includeLib ??= projectType === 'library';

  if (!schema.newProject) {
    const { buildTarget, serveTarget, unsuppored } =
      findExistingTargets(targets);

    /**
     * Here, we make sure that the project has a build target
     * and it is unsupported, before throwing the error.
     * The reason is that the project could not have a build
     * target at all, in which case we don't want to throw.
     * Or the project can have multiple build targets, one of which unsupported.
     * In that case, we convert the supported one.
     */
    if (!buildTarget && unsuppored) {
      throw new Error(
        `The project ${schema.project} cannot be converted to use the @nrwl/vite executors.`
      );
    }

    buildTargetName = buildTarget ?? buildTargetName;
    serveTargetName = serveTarget ?? serveTargetName;

    if (projectType === 'application') {
      moveAndEditIndexHtml(tree, schema, buildTargetName);
    }
    editTsConfig(tree, schema);
  }

  const initTask = await initGenerator(tree, {
    uiFramework: schema.uiFramework,
    includeLib: schema.includeLib,
  });
  tasks.push(initTask);

  addOrChangeBuildTarget(tree, schema, buildTargetName);

  if (!schema.includeLib) {
    addOrChangeServeTarget(tree, schema, serveTargetName);
  }

  writeViteConfig(tree, schema);

  if (schema.includeVitest) {
    const vitestTask = await vitestGenerator(tree, {
      project: schema.project,
      uiFramework: schema.uiFramework,
      inSourceTests: schema.inSourceTests,
      coverageProvider: 'c8',
      skipViteConfig: true,
    });
    tasks.push(vitestTask);
  }

  await formatFiles(tree);

  return runTasksInSerial(...tasks);
}

export default viteConfigurationGenerator;
export const configurationSchematic = convertNxGenerator(
  viteConfigurationGenerator
);
