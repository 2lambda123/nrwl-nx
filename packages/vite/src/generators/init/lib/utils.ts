import {
  addDependenciesToPackageJson,
  logger,
  readJson,
  readNxJson,
  Tree,
  updateJson,
  updateNxJson,
} from '@nx/devkit';

import {
  edgeRuntimeVmVersion,
  happyDomVersion,
  jsdomVersion,
  nxVersion,
  vitePluginDtsVersion,
  vitePluginReactSwcVersion,
  vitePluginReactVersion,
  vitestVersion,
  viteVersion,
} from '../../../utils/versions';
import { InitGeneratorSchema } from '../schema';

export function checkDependenciesInstalled(
  host: Tree,
  schema: InitGeneratorSchema
) {
  const packageJson = readJson(host, 'package.json');
  const devDependencies = {};
  const dependencies = {};
  packageJson.dependencies = packageJson.dependencies || {};
  packageJson.devDependencies = packageJson.devDependencies || {};

  // base deps
  devDependencies['@nx/vite'] = nxVersion;
  devDependencies['vite'] = viteVersion;
  devDependencies['vitest'] = vitestVersion;
  devDependencies['@vitest/ui'] = vitestVersion;

  if (schema.testEnvironment === 'jsdom') {
    devDependencies['jsdom'] = jsdomVersion;
  } else if (schema.testEnvironment === 'happy-dom') {
    devDependencies['happy-dom'] = happyDomVersion;
  } else if (schema.testEnvironment === 'edge-runtime') {
    devDependencies['@edge-runtime/vm'] = edgeRuntimeVmVersion;
  } else if (schema.testEnvironment !== 'node' && schema.testEnvironment) {
    logger.info(
      `A custom environment was provided: ${schema.testEnvironment}. You need to install it manually.`
    );
  }

  if (schema.uiFramework === 'react') {
    if (schema.compiler === 'swc') {
      devDependencies['@vitejs/plugin-react-swc'] = vitePluginReactSwcVersion;
    } else {
      devDependencies['@vitejs/plugin-react'] = vitePluginReactVersion;
    }
  }

  if (schema.includeLib) {
    devDependencies['vite-plugin-dts'] = vitePluginDtsVersion;
  }

  return addDependenciesToPackageJson(host, dependencies, devDependencies);
}

export function moveToDevDependencies(tree: Tree) {
  updateJson(tree, 'package.json', (packageJson) => {
    packageJson.dependencies = packageJson.dependencies || {};
    packageJson.devDependencies = packageJson.devDependencies || {};

    if (packageJson.dependencies['@nx/vite']) {
      packageJson.devDependencies['@nx/vite'] =
        packageJson.dependencies['@nx/vite'];
      delete packageJson.dependencies['@nx/vite'];
    }
    return packageJson;
  });
}

export function createVitestConfig(tree: Tree) {
  const nxJson = readNxJson(tree);

  const productionFileSet = nxJson.namedInputs?.production;
  if (productionFileSet) {
    productionFileSet.push(
      '!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)',
      '!{projectRoot}/tsconfig.spec.json'
    );

    nxJson.namedInputs.production = Array.from(new Set(productionFileSet));
  }

  nxJson.targetDefaults ??= {};
  nxJson.targetDefaults['@nx/vite:test'] ??= {};
  nxJson.targetDefaults['@nx/vite:test'].cache ??= true;
  nxJson.targetDefaults['@nx/vite:test'].inputs ??= [
    'default',
    productionFileSet ? '^production' : '^default',
  ];
  nxJson.targetDefaults['@nx/vite:test'].options ??= {
    passWithNoTests: true,
  };

  updateNxJson(tree, nxJson);
}

export function addPlugin(tree: Tree) {
  const nxJson = readNxJson(tree);
  nxJson.plugins ??= [];

  for (const plugin of nxJson.plugins) {
    if (
      typeof plugin === 'string'
        ? plugin === '@nx/vite/plugin'
        : plugin.plugin === '@nx/vite/plugin'
    ) {
      return;
    }
  }

  nxJson.plugins.push({
    plugin: '@nx/vite/plugin',
    options: {
      buildTargetName: 'build',
      previewTargetName: 'preview',
      testTargetName: 'test',
      serveTargetName: 'serve',
      serveStaticTargetName: 'serve-static',
    },
  });
  updateNxJson(tree, nxJson);
}
