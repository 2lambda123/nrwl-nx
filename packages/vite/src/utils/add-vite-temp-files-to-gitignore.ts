import type { Tree } from '@nx/devkit';

export function addViteTempFilesToGitIgnore(tree: Tree) {
  let gitIgnoreContents = `**/vite.config.{js,ts,mjs,mts,cjs,cts}.timestamp*`;
  if (tree.exists('.gitignore')) {
    gitIgnoreContents = `${tree.read('.gitignore', 'utf-8')}
    ${gitIgnoreContents}`;
  }

  tree.write('.gitignore', gitIgnoreContents);
}
