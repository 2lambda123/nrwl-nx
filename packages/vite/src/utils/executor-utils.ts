import { printDiagnostics, runTypeCheck } from '@nx/js';
import { join, resolve } from 'path';
import { ViteBuildExecutorOptions } from '../executors/build/schema';
import { ExecutorContext } from '@nx/devkit';
import { ViteDevServerExecutorOptions } from '../executors/dev-server/schema';
import {
  calculateProjectDependencies,
  createTmpTsConfig,
} from '@nx/js/src/utils/buildable-libs-utils';

export async function validateTypes(opts: {
  workspaceRoot: string;
  projectRoot: string;
  tsconfig: string;
}): Promise<void> {
  const result = await runTypeCheck({
    workspaceRoot: opts.workspaceRoot,
    tsConfigPath: join(opts.workspaceRoot, opts.tsconfig),
    mode: 'noEmit',
  });

  await printDiagnostics(result.errors, result.warnings);

  if (result.errors.length > 0) {
    throw new Error('Found type errors. See above.');
  }
}

export function createBuildableTsConfig(
  projectRoot: string,
  options: ViteBuildExecutorOptions | ViteDevServerExecutorOptions,
  context: ExecutorContext
) {
  const tsConfig = resolve(projectRoot, 'tsconfig.json');
  options.buildLibsFromSource ??= true;

  if (!options.buildLibsFromSource) {
    const { dependencies } = calculateProjectDependencies(
      context.projectGraph,
      context.root,
      context.projectName,
      context.targetName,
      context.configurationName
    );
    // this tsconfig is used via the vite ts paths plugin
    createTmpTsConfig(tsConfig, context.root, projectRoot, dependencies);
  }
}
