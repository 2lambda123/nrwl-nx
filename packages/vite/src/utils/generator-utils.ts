import {
  joinPathFragments,
  logger,
  offsetFromRoot,
  readJson,
  readProjectConfiguration,
  TargetConfiguration,
  Tree,
  updateProjectConfiguration,
  writeJson,
} from '@nrwl/devkit';
import { ViteBuildExecutorOptions } from '../executors/build/schema';
import { ViteDevServerExecutorOptions } from '../executors/dev-server/schema';
import { VitestExecutorOptions } from '../executors/test/schema';
import { Schema } from '../generators/configuration/schema';

/**
 * This function is used to find the build and serve targets for
 * an application.
 *
 * The reason this function exists is because we cannot assume
 * that the user has not created a custom build target for the application,
 * or that they have not changed the name of the build target
 * from build to anything else.
 *
 * So, in order to find the correct name of the target,
 * we have to look into all the targets, check the executor
 * they are using, and infer from the executor that the target
 * is a build target.
 */
export function findExistingTargets(targets: {
  [targetName: string]: TargetConfiguration;
}): {
  buildTarget: string;
  serveTarget: string;
  testTarget: string;
} {
  const returnObject: {
    buildTarget: string;
    serveTarget: string;
    testTarget: string;
  } = {
    buildTarget: 'build',
    serveTarget: 'serve',
    testTarget: 'test',
  };

  Object.entries(targets).forEach(([target, targetConfig]) => {
    switch (targetConfig.executor) {
      case '@nrwl/webpack:dev-server':
      case '@nxext/vite:dev':
        returnObject.serveTarget = target;
        break;
      case '@angular-devkit/build-angular:browser':
        /**
         * Not looking for '@nrwl/angular:ng-packagr-lite' or any other
         * angular executors.
         * Only looking for '@angular-devkit/build-angular:browser'
         * because the '@nrwl/angular:ng-packagr-lite' executor
         * (and maybe the other custom exeucutors) is only used for libs atm.
         */
        returnObject.buildTarget = target;
        break;
      case '@nrwl/webpack:webpack':
      case '@nrwl/next:build':
      case '@nrwl/web:webpack':
      case '@nrwl/web:rollup':
      case '@nrwl/js:tsc':
      case '@nrwl/angular:ng-packagr-lite':
      case '@nrwl/js:babel':
      case '@nrwl/js:swc':
      case '@nxext/vite:build':
        returnObject.buildTarget = target;
        break;
      case '@nrwl/jest:jest':
      case 'nxext/vitest:vitest':
        returnObject.testTarget = target;
      default:
        returnObject.buildTarget = 'build';
        returnObject.serveTarget = 'serve';
        returnObject.testTarget = 'test';
        break;
    }
  });

  return returnObject;
}

export function addOrChangeTestTarget(
  tree: Tree,
  options: Schema,
  target: string
) {
  const project = readProjectConfiguration(tree, options.project);
  const targets = {
    ...project.targets,
  };

  const testOptions: VitestExecutorOptions = {
    passWithNoTests: true,
  };

  if (targets[target]) {
    targets[target].executor = '@nrwl/vite:test';
    delete targets[target].options.jestConfig;
  } else {
    targets[target] = {
      executor: '@nrwl/vite:test',
      outputs: ['{projectRoot}/coverage'],
      options: testOptions,
    };
  }

  updateProjectConfiguration(tree, options.project, {
    ...project,
    targets: {
      ...targets,
    },
  });
}

export function addOrChangeBuildTarget(
  tree: Tree,
  options: Schema,
  target: string
) {
  const project = readProjectConfiguration(tree, options.project);

  const buildOptions: ViteBuildExecutorOptions = {
    outputPath: joinPathFragments(
      'dist',
      project.root != '.' ? project.root : options.project
    ),
  };

  const targets = {
    ...project.targets,
  };

  if (targets[target]) {
    targets[target].options = {
      ...buildOptions,
    };
    targets[target].executor = '@nrwl/vite:build';
  } else {
    targets[`${target}`] = {
      executor: '@nrwl/vite:build',
      outputs: ['{options.outputPath}'],
      defaultConfiguration: 'production',
      options: buildOptions,
      configurations: {
        development: {},
        production: {
          fileReplacements: [
            {
              replace: joinPathFragments(
                project.sourceRoot,
                'environments/environment.ts'
              ),
              with: joinPathFragments(
                project.sourceRoot,
                'environments/environment.prod.ts'
              ),
            },
          ],
          sourceMap: false,
        },
      },
    };
  }

  updateProjectConfiguration(tree, options.project, {
    ...project,
    targets: {
      ...targets,
    },
  });
}

export function addOrChangeServeTarget(
  tree: Tree,
  options: Schema,
  target: string
) {
  const project = readProjectConfiguration(tree, options.project);

  const serveOptions: ViteDevServerExecutorOptions = {
    buildTarget: `${options.project}:build`,
  };

  const targets = {
    ...project.targets,
  };

  if (targets[target]) {
    targets[target].options = {
      ...serveOptions,
    };
    targets[target].executor = '@nrwl/vite:dev-server';
  } else {
    targets[`${target}`] = {
      executor: '@nrwl/vite:dev-server',
      defaultConfiguration: 'development',
      options: {
        buildTarget: `${options.project}:build`,
      },
      configurations: {
        development: {
          buildTarget: `${options.project}:build:development`,
        },
        production: {
          buildTarget: `${options.project}:build:production`,
          hmr: false,
        },
      },
    };
  }

  updateProjectConfiguration(tree, options.project, {
    ...project,
    targets: {
      ...targets,
    },
  });
}

export function editTsConfig(tree: Tree, options: Schema) {
  const projectConfig = readProjectConfiguration(tree, options.project);

  const config = readJson(tree, `${projectConfig.root}/tsconfig.json`);

  switch (options.uiFramework) {
    case 'react':
      config.compilerOptions = {
        target: 'ESNext',
        useDefineForClassFields: true,
        lib: ['DOM', 'DOM.Iterable', 'ESNext'],
        allowJs: false,
        skipLibCheck: true,
        esModuleInterop: false,
        allowSyntheticDefaultImports: true,
        strict: true,
        forceConsistentCasingInFileNames: true,
        module: 'ESNext',
        moduleResolution: 'Node',
        resolveJsonModule: true,
        isolatedModules: true,
        noEmit: true,
        jsx: 'react-jsx',
        types: ['vite/client'],
      };
      config.include = [...config.include, 'src'];
      break;
    case 'none':
      config.compilerOptions = {
        target: 'ESNext',
        useDefineForClassFields: true,
        module: 'ESNext',
        lib: ['ESNext', 'DOM'],
        moduleResolution: 'Node',
        strict: true,
        resolveJsonModule: true,
        isolatedModules: true,
        esModuleInterop: true,
        noEmit: true,
        noUnusedLocals: true,
        noUnusedParameters: true,
        noImplicitReturns: true,
        skipLibCheck: true,
        types: ['vite/client'],
      };
      config.include = [...config.include, 'src'];
      break;
    default:
      break;
  }

  writeJson(tree, `${projectConfig.root}/tsconfig.json`, config);
}

export function moveAndEditIndexHtml(
  tree: Tree,
  options: Schema,
  buildTarget: string
) {
  const projectConfig = readProjectConfiguration(tree, options.project);

  let indexHtmlPath =
    projectConfig.targets[buildTarget].options?.index ??
    `${projectConfig.root}/src/index.html`;
  const mainPath = (
    projectConfig.targets[buildTarget].options?.main ??
    `${projectConfig.root}/src/main.ts${
      options.uiFramework === 'react' ? 'x' : ''
    }`
  ).replace(projectConfig.root, '');

  if (
    !tree.exists(indexHtmlPath) &&
    tree.exists(`${projectConfig.root}/index.html`)
  ) {
    indexHtmlPath = `${projectConfig.root}/index.html`;
  }

  if (tree.exists(indexHtmlPath)) {
    const indexHtmlContent = tree.read(indexHtmlPath, 'utf8');
    if (
      !indexHtmlContent.includes(
        `<script type="module" src="${mainPath}"></script>`
      )
    ) {
      tree.write(
        `${projectConfig.root}/index.html`,
        indexHtmlContent.replace(
          '</body>',
          `<script type="module" src="${mainPath}"></script>
          </body>`
        )
      );

      if (tree.exists(`${projectConfig.root}/src/index.html`))
        tree.delete(`${projectConfig.root}/src/index.html`);
    }
  } else {
    tree.write(
      `${projectConfig.root}/index.html`,
      `<!DOCTYPE html>
      <html lang="en">
        <head>
          <meta charset="UTF-8" />
          <link rel="icon" href="/favicon.ico" />
          <meta name="viewport" content="width=device-width, initial-scale=1.0" />
          <title>Vite</title>
        </head>
        <body>
          <div id="root"></div>
          <script type="module" src="${mainPath}"></script>
        </body>
      </html>`
    );
  }
}

export function writeViteConfig(tree: Tree, options: Schema) {
  const projectConfig = readProjectConfiguration(tree, options.project);

  const viteConfigPath = `${projectConfig.root}/vite.config.ts`;

  if (tree.exists(viteConfigPath)) {
    // TODO (katerina): Ideally we should check if the config is already set up correctly
    logger.info(
      `vite.config.ts already exists. Skipping creation of vite config for ${options.project}.`
    );
    return;
  }

  let viteConfigContent = '';

  const testOption = `test: {
    globals: true,
    environment: 'jsdom',
    include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
    ${
      options.inSourceTests
        ? `includeSource: ['src/**/*.{js,mjs,cjs,ts,mts,cts,jsx,tsx}']`
        : ''
    }
  },`;

  const defineOption = `define: {
    'import.meta.vitest': undefined
  },`;

  switch (options.uiFramework) {
    case 'react':
      viteConfigContent = `
${options.includeVitest ? '/// <reference types="vitest" />' : ''}
      import { defineConfig } from 'vite';
      import react from '@vitejs/plugin-react';
      import ViteTsConfigPathsPlugin from 'vite-tsconfig-paths';
      
      export default defineConfig({
        plugins: [
          react(),
          ViteTsConfigPathsPlugin({
            root: '${offsetFromRoot(projectConfig.root)}',
            projects: ['tsconfig.base.json'],
          }),
        ],
        ${options.inSourceTests ? defineOption : ''}
        ${options.includeVitest ? testOption : ''}
      });`;
      break;
    case 'none':
      viteConfigContent = `
      ${options.includeVitest ? '/// <reference types="vitest" />' : ''}
      import { defineConfig } from 'vite';
      import ViteTsConfigPathsPlugin from 'vite-tsconfig-paths';
      
      export default defineConfig({
        plugins: [
          ViteTsConfigPathsPlugin({
            root: '${offsetFromRoot(projectConfig.root)}',
            projects: ['tsconfig.base.json'],
          }),
        ],
        ${options.inSourceTests ? defineOption : ''}
        ${options.includeVitest ? testOption : ''}
      });`;
      break;
    default:
      break;
  }

  tree.write(viteConfigPath, viteConfigContent);
}
