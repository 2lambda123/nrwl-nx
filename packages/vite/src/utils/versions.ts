export const nxVersion = require('../../package.json').version;
export const viteVersion = '^3.0.5';
export const vitePluginEslintVersion = '^1.6.0';
export const vitestVersion = '^0.19.1';
export const vitestUiVersion = '^0.9.3';
export const vitePluginReactVersion = '^2.2.0';
export const vitePluginVueVersion = '^3.2.0';
export const vitePluginVueJsxVersion = '^2.1.1';
export const viteTsConfigPathsVersion = '^3.5.2';
