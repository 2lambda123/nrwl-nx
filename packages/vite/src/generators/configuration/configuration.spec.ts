import { addDependenciesToPackageJson, readJson, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyV1Workspace } from '@nrwl/devkit/testing';
import { nxVersion } from '../../utils/versions';

import { viteConfigurationGenerator } from './configuration';
import {
  mockReactAppGenerator,
  mockWebAppGenerator,
} from '../../utils/test-utils';

describe('@nrwl/vite:configuration', () => {
  let tree: Tree;

  describe('transform React app to use Vite', () => {
    beforeAll(async () => {
      tree = createTreeWithEmptyV1Workspace();
      await mockReactAppGenerator(tree);
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nrwl/vite': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await viteConfigurationGenerator(tree, {
        uiFramework: 'react',
        project: 'my-test-react-app',
      });
    });
    it('should add vite packages and react-related dependencies for vite', async () => {
      const packageJson = readJson(tree, '/package.json');
      expect(packageJson.devDependencies).toMatchObject({
        vite: expect.any(String),
        '@vitejs/plugin-react': expect.any(String),
      });
    });

    it('should move index.html to the root of the project', () => {
      expect(tree.exists('apps/my-test-react-app/src/index.html')).toBeFalsy();
      expect(tree.exists('apps/my-test-react-app/index.html')).toBeTruthy();
    });

    it('should create correct tsconfig compilerOptions', () => {
      const tsconfigJson = readJson(
        tree,
        'apps/my-test-react-app/tsconfig.json'
      );
      expect(tsconfigJson.compilerOptions.types).toMatchObject(['vite/client']);
    });

    it('should create vite.config file at the root of the app', () => {
      expect(tree.exists('apps/my-test-react-app/vite.config.ts')).toBe(true);
    });
  });

  describe('transform Web app to use Vite', () => {
    beforeAll(async () => {
      tree = createTreeWithEmptyV1Workspace();
      await mockWebAppGenerator(tree);
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nrwl/vite': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await viteConfigurationGenerator(tree, {
        uiFramework: 'none',
        project: 'my-test-web-app',
      });
    });
    it('should add vite dependencies for vite', async () => {
      const packageJson = readJson(tree, '/package.json');
      expect(packageJson.devDependencies).toMatchObject({
        vite: expect.any(String),
        'vite-tsconfig-paths': expect.any(String),
      });
    });

    it('should move index.html to the root of the project', () => {
      expect(tree.exists('apps/my-test-web-app/src/index.html')).toBeFalsy();
      expect(tree.exists('apps/my-test-web-app/index.html')).toBeTruthy();
    });

    it('should create correct tsconfig compilerOptions', () => {
      const tsconfigJson = readJson(tree, 'apps/my-test-web-app/tsconfig.json');
      expect(tsconfigJson.compilerOptions.types).toMatchObject(['vite/client']);
    });

    it('should create vite.config file at the root of the app', () => {
      expect(tree.exists('apps/my-test-web-app/vite.config.ts')).toBe(true);
    });
  });

  describe('vitest', () => {
    beforeAll(async () => {
      tree = createTreeWithEmptyV1Workspace();
      await mockReactAppGenerator(tree);
      const existing = 'existing';
      const existingVersion = '1.0.0';
      addDependenciesToPackageJson(
        tree,
        { '@nrwl/vite': nxVersion, [existing]: existingVersion },
        { [existing]: existingVersion }
      );
      await viteConfigurationGenerator(tree, {
        uiFramework: 'react',
        project: 'my-test-react-app',
        includeVitest: true,
      });
    });
    it('should create a vitest configuration if "includeVitest" is true', () => {
      const viteConfig = tree
        .read('apps/my-test-react-app/vite.config.ts')
        .toString();
      console.log(viteConfig);
      expect(viteConfig).toContain('test');
    });
  });
});
