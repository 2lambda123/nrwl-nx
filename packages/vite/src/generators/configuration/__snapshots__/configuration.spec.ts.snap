// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nrwl/vite:configuration transform React app to use Vite should transform workspace.json project config 1`] = `
"{
  \\"projects\\": {
    \\"my-test-react-app\\": {
      \\"$schema\\": \\"../../node_modules/nx/schemas/project-schema.json\\",
      \\"root\\": \\"apps/my-test-react-app\\",
      \\"sourceRoot\\": \\"apps/my-test-react-app/src\\",
      \\"projectType\\": \\"application\\",
      \\"architect\\": {
        \\"build\\": {
          \\"builder\\": \\"@nrwl/vite:build\\",
          \\"outputs\\": [
            \\"{options.outputPath}\\"
          ],
          \\"defaultConfiguration\\": \\"production\\",
          \\"options\\": {
            \\"outputPath\\": \\"dist/apps/my-test-react-app\\"
          },
          \\"configurations\\": {
            \\"development\\": {
              \\"extractLicenses\\": false,
              \\"optimization\\": false,
              \\"sourceMap\\": true,
              \\"vendorChunk\\": true
            },
            \\"production\\": {
              \\"fileReplacements\\": [
                {
                  \\"replace\\": \\"apps/my-test-react-app/src/environments/environment.ts\\",
                  \\"with\\": \\"apps/my-test-react-app/src/environments/environment.prod.ts\\"
                }
              ],
              \\"optimization\\": true,
              \\"outputHashing\\": \\"all\\",
              \\"sourceMap\\": false,
              \\"namedChunks\\": false,
              \\"extractLicenses\\": true,
              \\"vendorChunk\\": false
            }
          }
        },
        \\"serve\\": {
          \\"builder\\": \\"@nrwl/vite:dev-server\\",
          \\"defaultConfiguration\\": \\"development\\",
          \\"options\\": {
            \\"buildTarget\\": \\"my-test-react-app:build\\"
          },
          \\"configurations\\": {
            \\"development\\": {
              \\"buildTarget\\": \\"my-test-react-app:build:development\\"
            },
            \\"production\\": {
              \\"buildTarget\\": \\"my-test-react-app:build:production\\",
              \\"hmr\\": false
            }
          }
        },
        \\"lint\\": {
          \\"builder\\": \\"@nrwl/linter:eslint\\",
          \\"outputs\\": [
            \\"{options.outputFile}\\"
          ],
          \\"options\\": {
            \\"lintFilePatterns\\": [
              \\"apps/my-test-react-app/**/*.{ts,tsx,js,jsx}\\"
            ]
          }
        },
        \\"test\\": {
          \\"builder\\": \\"@nrwl/jest:jest\\",
          \\"outputs\\": [
            \\"{workspaceRoot}/coverage/{projectRoot}\\"
          ],
          \\"options\\": {
            \\"jestConfig\\": \\"apps/my-test-react-app/jest.config.ts\\",
            \\"passWithNoTests\\": true
          }
        }
      },
      \\"tags\\": []
    }
  },
  \\"version\\": 1
}
"
`;

exports[`@nrwl/vite:configuration transform Web app to use Vite should transform workspace.json project config 1`] = `
"{
  \\"projects\\": {
    \\"my-test-web-app\\": {
      \\"$schema\\": \\"../../node_modules/nx/schemas/project-schema.json\\",
      \\"projectType\\": \\"application\\",
      \\"root\\": \\"apps/my-test-web-app\\",
      \\"sourceRoot\\": \\"apps/my-test-web-app/src\\",
      \\"tags\\": [],
      \\"architect\\": {
        \\"build\\": {
          \\"builder\\": \\"@nrwl/vite:build\\",
          \\"outputs\\": [
            \\"{options.outputPath}\\"
          ],
          \\"defaultConfiguration\\": \\"production\\",
          \\"options\\": {
            \\"outputPath\\": \\"dist/apps/my-test-web-app\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"optimization\\": true,
              \\"outputHashing\\": \\"all\\",
              \\"sourceMap\\": false,
              \\"namedChunks\\": false,
              \\"extractLicenses\\": true,
              \\"vendorChunk\\": false,
              \\"fileReplacements\\": [
                {
                  \\"replace\\": \\"apps/my-test-web-app/src/environments/environment.ts\\",
                  \\"with\\": \\"apps/my-test-web-app/src/environments/environment.prod.ts\\"
                }
              ]
            }
          }
        },
        \\"serve\\": {
          \\"builder\\": \\"@nrwl/vite:dev-server\\",
          \\"options\\": {
            \\"buildTarget\\": \\"my-test-web-app:build\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"buildTarget\\": \\"my-test-web-app:build:production\\"
            }
          }
        },
        \\"lint\\": {
          \\"builder\\": \\"@nrwl/linter:eslint\\",
          \\"outputs\\": [
            \\"{options.outputFile}\\"
          ],
          \\"options\\": {
            \\"lintFilePatterns\\": [
              \\"apps/my-test-web-app/**/*.ts\\"
            ]
          }
        },
        \\"test\\": {
          \\"builder\\": \\"@nrwl/jest:jest\\",
          \\"outputs\\": [
            \\"{workspaceRoot}/coverage/{projectRoot}\\"
          ],
          \\"options\\": {
            \\"jestConfig\\": \\"apps/my-test-web-app/jest.config.ts\\",
            \\"passWithNoTests\\": true
          }
        }
      }
    }
  },
  \\"version\\": 1
}
"
`;
